### ISSUE TITLE
Cannot use invalid module set default

### ISSUE THREAD

**Issue by iskra-anl** [LLNL/spack#37649]
### Steps to reproduce

Are the modules currently broken? I installed a new version of a package and noticed that the output of `module avail` did not get updated. This was on OLCF Crusher, BTW. I tried uninstalling/reinstalling, deleting existing files from `$spack/share/spack/modules/`... No matter what I try, no new files are generated there. I tried manually:

```console
$ spack module tcl refresh
==> Error: Cannot use invalid module set default.    Valid module set names are []
```

### Error message

```console
$ spack --debug --stacktrace module tcl refresh
autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/cmd/__init__.py:123 ==> [2023-05-12-13:15:32.554971] Imported module from built-in commands
autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/cmd/__init__.py:123 ==> [2023-05-12-13:15:32.558873] Imported module from built-in commands
autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/config.py:996 ==> [2023-05-12-13:15:32.612730] Reading config from file /autofs/nccs-svm1_home1/iskra/spack/etc/spack/defaults/config.yaml
autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/database.py:391 ==> [2023-05-12-13:15:32.664261] DATABASE LOCK TIMEOUT: 60s
autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/database.py:397 ==> [2023-05-12-13:15:32.664440] PACKAGE LOCK TIMEOUT: No timeout
autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/config.py:996 ==> [2023-05-12-13:15:32.685620] Reading config from file /autofs/nccs-svm1_home1/iskra/spack/etc/spack/defaults/repos.yaml
autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/config.py:996 ==> [2023-05-12-13:15:32.822047] Reading config from file /autofs/nccs-svm1_home1/iskra/spack/etc/spack/defaults/modules.yaml
autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/config.py:996 ==> [2023-05-12-13:15:32.841633] Reading config from file /autofs/nccs-svm1_home1/iskra/spack/etc/spack/defaults/linux/modules.yaml
autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/main.py:1021 ==> [2023-05-12-13:15:32.845196] ConfigError: Cannot use invalid module set default.    Valid module set names are []
autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/error.py:54 ==> [2023-05-12-13:15:32.846711] Error: Cannot use invalid module set default.    Valid module set names are []
Traceback (most recent call last):
  File "/autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/main.py", line 1018, in main
    return _main(argv)
  File "/autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/main.py", line 973, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/main.py", line 1001, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/main.py", line 650, in _invoke_command
    return_val = command(parser, args)
  File "/autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/cmd/module.py", line 26, in module
    _subcommands[args.module_command](parser, args)
  File "/autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/cmd/modules/__init__.py", line 376, in modules_cmd
    callbacks[args.subparser_name](module_type, specs, args)
  File "/autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/cmd/modules/__init__.py", line 284, in refresh
    check_module_set_name(args.module_set_name)
  File "/autofs/nccs-svm1_home1/iskra/spack/lib/spack/spack/cmd/modules/__init__.py", line 133, in check_module_set_name
    raise spack.config.ConfigError(msg)
spack.config.ConfigError: Cannot use invalid module set default.    Valid module set names are []
```

### Information on your system

* **Spack:** 0.20.0.dev0 (e2ae60a3b0535c553f6c5a13e7106abbbeef36cb)
* **Python:** 3.6.15
* **Platform:** linux-sles15-zen3
* **Concretizer:** clingo

This is on the OLCF Crusher

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by vmiheer** [comment]
FYI: 99c3ecc139 works fine.

**Comment by vmiheer** [comment]
99c3ecc139 works okay.

**Comment by amir-raoofy** [comment]
I am observing the exact same issue (Beast test system at LRZ)

```
$ spack --debug --stacktrace module lmod refresh --delete-tree -y
lib/spack/spack/cmd/__init__.py:123 ==> [2023-05-15-12:33:01.566495] Imported module from built-in commands
lib/spack/spack/cmd/__init__.py:123 ==> [2023-05-15-12:33:01.584486] Imported module from built-in commands
lib/spack/spack/config.py:996 ==> [2023-05-15-12:33:01.591875] Reading config from file /home/sw/spack/etc/spack/defaults/config.yaml
lib/spack/spack/database.py:391 ==> [2023-05-15-12:33:01.632480] DATABASE LOCK TIMEOUT: 60s
lib/spack/spack/database.py:397 ==> [2023-05-15-12:33:01.632721] PACKAGE LOCK TIMEOUT: No timeout
lib/spack/spack/config.py:996 ==> [2023-05-15-12:33:01.645928] Reading config from file /home/sw/spack/etc/spack/defaults/repos.yaml
lib/spack/spack/config.py:996 ==> [2023-05-15-12:33:01.671941] Reading config from file /home/sw/spack/etc/spack/defaults/modules.yaml
lib/spack/spack/config.py:996 ==> [2023-05-15-12:33:01.694013] Reading config from file /home/sw/spack/etc/spack/defaults/linux/modules.yaml
lib/spack/spack/main.py:1021 ==> [2023-05-15-12:33:01.697672] ConfigError: Cannot use invalid module set default.    Valid module set names are []
lib/spack/spack/error.py:54 ==> [2023-05-15-12:33:01.697945] Error: Cannot use invalid module set default.    Valid module set names are []
Traceback (most recent call last):
  File "/home/sw/spack/lib/spack/spack/main.py", line 1018, in main
    return _main(argv)
  File "/home/sw/spack/lib/spack/spack/main.py", line 973, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/home/sw/spack/lib/spack/spack/main.py", line 1001, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/home/sw/spack/lib/spack/spack/main.py", line 650, in _invoke_command
    return_val = command(parser, args)
  File "/home/sw/spack/lib/spack/spack/cmd/module.py", line 26, in module
    _subcommands[args.module_command](parser, args)
  File "/home/sw/spack/lib/spack/spack/cmd/modules/__init__.py", line 376, in modules_cmd
    callbacks[args.subparser_name](module_type, specs, args)
  File "/home/sw/spack/lib/spack/spack/cmd/modules/__init__.py", line 284, in refresh
    check_module_set_name(args.module_set_name)
  File "/home/sw/spack/lib/spack/spack/cmd/modules/__init__.py", line 133, in check_module_set_name
    raise spack.config.ConfigError(msg)
spack.config.ConfigError: Cannot use invalid module set default.    Valid module set names are []
di75gix@sap11:/home/sw/spack>
```

* **Spack:** 0.20.0.dev0 (e1344b5497e8f9af5c42020a4cf12e62d6112844)
* **Python:** 3.6.15
* **Platform:** linux-sles15-icelake
* **Concretizer:** clingo

**Comment by mgjf** [comment]
FYI: I also get this error in `releases/v0.20`

**Comment by haampie** [comment]
Can you share `spack config blame modules`?

The logic is:

```python
valid_names = set(
        [
            key
            for key, value in modules_config.items()
            if isinstance(value, dict) and value.get("enable", [])
        ]
    )
```

which is apparently empty for you all

**Comment by haampie** [comment]
Nevermind, the issue is that tcl modules are disabled by default now, which you should get warnings for if you have custom tcl config.

To re-enable tcl module generation:

```
spack config add 'modules:default:enable:[tcl]'
```

I admit the error you get is not very informative

**Comment by amir-raoofy** [comment]
This solution works for me; thanks!

**Comment by haampie** [comment]
Iteration 2:

I believe `modules:name:enable:[...]` doesn't necessary have to list `tcl` for `spack module tcl refresh` to work, it merely enables auto-generation of modules post installation.

So... it was indeed a bug, and no config should be changed for anybody.

https://github.com/spack/spack/pull/37701 should fix this.

**Comment by amir-raoofy** [comment]
Yes, It was not necessary to manually list tcl before (v0.19.2).

**Comment by haampie** [comment]
Should be fixed now with #37701
